### ISSUE TITLE
Any way to add arguments of evaluation functions conditionally depending on the config/cli arguments?

### ISSUE THREAD

**Issue by florianblume** [omni-us/jsonargparse#119]
I have a setup where I have multiple evaluation functions in modules in a folder called `evaluation`. Now, I'd like to give the user (essentially only myself ;) ) the option in the configs/CLI arguments to specify which module and which function they'd like to execute. The functions share some arguments but some are function-specific. When I add the function arguments to the parser it of course always adds them and I cannot do something like `eval_func(**config['eval']['params'])`.

Is there any way I'm not seeing to conditionally add such function arguments without having to manually scan the CLI arguments/config files to see which eval function is to be executed?

I'm open to any simple solution, doesn't have to be the way I'm trying to do it right now (e.g. if e.g. creating classes comes to the rescue so be it but I don't see how that could work). Any help is appreciated :)

**Comment by mauvilsa** [comment]
The subcommands work conditionally. Each subcommand that you add to a main parser will have its own independent parser with whatever arguments it should have. You could write the code that builds this parser with the subcommands, but from what I understand https://jsonargparse.readthedocs.io/en/stable/#jsonargparse.CLI would fit your needs perfectly. When you give a class as a component to the CLI, all its methods become subcommands, and when the CLI is run the selected method will be run. Since you already have the functions you can create a class to group them. The following gives the idea:

```python
from jsonargparse import CLI

def foo(foo1: str, foo2: int = 1):
    """This is foo

    Args:
        foo1: First param
        foo2: Second param
    """
    print(f'foo called with foo1={foo1} and foo2={foo2}')

def bar(bar1: int = 2, bar2: int = 3):
    """This is bar

    Args:
        bar1: First param
        bar2: Second param
    """
    print(f'bar called with bar1={bar1} and bar2={bar2}')

class Subcommands:
    foo = staticmethod(foo)
    bar = staticmethod(bar)

CLI(Subcommands)
```

Running this would be like:

```bash
$ pythonr cli.py foo hello
foo called with foo1=hello and foo2=1
$ python cli.py bar --bar2 9
bar called with bar1=2 and bar2=9
```

Since you have several modules, each could have its own class. When multiple classes are given as components, i.e. `CLI([SubcomsModule1, SubcomsModule2])`, then there will be two levels of subcommands. Thus to run would be like `python cli.py SubcomsModule2 eval_func4 ...`.

**Comment by florianblume** [comment]
Hey @mauvilsa, thanks for the quick reply!

I'm still trying to work out how to convert that to my case. I'm actually using Pytorch Lightning and its CLI which is a class and not a function like here. I tried to use the subcommand structure already and now again with your advice by creating a class for the individual eval methods but had no success yet.

One issue is that I'm processing commands differently than in standard Pytorch Lightning, as PL only allows to execute one command but I need cascades of commands (like train->test, etc.). That's why I added a commands arg which accepts command strings like `train-test`. `eval` will probably always be called on its own so I probably could combine the two concepts anyway and modify the sys args when command is set to `eval` but I'm not sure if this will work.

I'll try to think about this some more, thanks for the help again!
